lib/db/lowlevel.go

```diff
@@ -826,16 +832,31 @@ func (db *Lowlevel) gcIndirect(ctx context.Context) (err error) {
 		return err
 	}
 
-	l.Debugf("Finished GC, starting compaction (discarded/remaining: %v/%v blocks, %v/%v versions)", discardedBlocks, matchedBlocks, discardedVersions, matchedVersions)
+	l.Debugf("Finished GC (discarded/remaining: %v/%v blocks, %v/%v versions)", discardedBlocks, matchedBlocks, discardedVersions, matchedVersions)
+
+	return nil
+}
+
+func (db *Lowlevel) recordIndirectionHashesForFile(f *protocol.FileInfo) {
+	db.recordIndirectionHashes(IndirectionHashesOnly{BlocksHash: f.BlocksHash, VersionHash: f.VersionHash})
+}
 
-	return db.Compact()
+func (db *Lowlevel) recordIndirectionHashes(hs IndirectionHashesOnly) {
+	db.filtersMut.Lock()
```

Comment:
As the lock only protects concurrent access to the blockfilters, but not the blockfilter variables (that's `gcMut`), we could first do the nil checks and abort early, thus preventing pointless locking.

Reply:
Blobloom has a SyncFilter type that uses atomic operations for Add and Has. It can be used to omit this lock entirely.

Reply:
@imsodin Right!

@greatroar Thanks, I wasn't aware. In this case though we really only need the locking on the Add() operation.

Reply:
Looking at it I see it's probably more efficient to use a SyncFilter anyway. Doing that instead. 👍